components/brave_federated/communication_adapter.cc

```diff
@@ -0,0 +1,186 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_federated/communication_adapter.h"
+
+#include <utility>
+#include <vector>
+
+#include "base/logging.h"
+#include "base/memory/raw_ptr.h"
+#include "brave/components/brave_federated/adapters/flower_helper.h"
+#include "brave/components/brave_federated/features.h"
+#include "brave/components/brave_federated/task/typing.h"
+#include "net/http/http_request_headers.h"
+#include "net/http/http_status_code.h"
+#include "net/traffic_annotation/network_traffic_annotation.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+#include "url/gurl.h"
+
+namespace brave_federated {
+
+namespace {
+
+// Maximum size of the federated server response in bytes.
+const int kMaxFederatedServerResponseSizeBytes = 1024 * 1024;  // 1 MB
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("federated_learning", R"(
+        semantics {
+          sender: "Federated Learning"
+          description:
+            "Federated model updates based off toy task and locally generated "
+            "synthetic dataset."
+          trigger:
+            "Reports are generated when the brave-federated flag is enabled "
+            "and learning tasks are made available by the federated server. "
+            "Reports are NOT generated when the user is on battery power "
+            "or when the device is on a metered network (not WiFi/Ethernet). "
+          data:
+            "Simple federated model updates based off toy task and locally "
+            "generated synthetic data."
+          destination: WEBSITE
+        }
+        policy {
+          cookies_allowed: NO
+          setting:
+            "This experimental feature is only enabled in Nightly builds and"
+            "can be disabled via the 'brave://flags/#brave-federated' "
+          policy_exception_justification:
+            "Not implemented."
+        }
+    )");
+}
+
+}  // namespace
+
+CommunicationAdapter::CommunicationAdapter(
```

Comment:
this class should have tests

Reply:
This is only handling the network part of the requests. The parsing/payload creation is carried out in `flower_helper` which we have added tests for.